src/wallet/rpcdump.cpp

```diff
@@ -892,180 +887,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            }
 
-                CTxDestination pubkey_dest = pubKey.GetID();
+        // P2WSH
+        } else if (!strWitnessScript.empty() && script.IsPayToWitnessScriptHash()) {
+            if (!IsHex(strWitnessScript)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid witness script");
+            }
 
-                // Consistency check.
-                if (!(pubkey_dest == dest)) {
-                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Consistency check failed");
-                }
+            // Generate the scripts
+            std::vector<unsigned char> vData(ParseHex(strWitnessScript));
+            CScript witnessScript = CScript(vData.begin(), vData.end());
+            CScriptID witness_id(witnessScript);
+            CScript witnessDestination = GetScriptForDestination(WitnessV0ScriptHash(witnessScript));
 
-                CScript pubKeyScript = GetScriptForDestination(pubkey_dest);
+            // Check that the witnessScript and scriptPubKey match
+            if (witnessDestination != script) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "The witnessScript does not match the scriptPubKey");
+            }
 
-                if (::IsMine(*pwallet, pubKeyScript) == ISMINE_SPENDABLE) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
-                }
+            // Import into the wallet
+            if (!pwallet->AddWatchOnly(witnessScript, timestamp)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+            }
 
-                pwallet->MarkDirty();
+            if (!pwallet->HaveCScript(witness_id) && !pwallet->AddCScript(witnessScript)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2wsh witnessScript to wallet");
+            }
 
-                if (!pwallet->AddWatchOnly(pubKeyScript, timestamp)) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+        // P2PK/P2PKH/P2WPKH
+        } else if (dest.type() == typeid(CKeyID) || dest.type() == typeid(WitnessV0KeyHash)) {
+            if (keys.size() > 1 || pubKeys.size() > 1) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "More than private key given for one address");
+            }
+            CPubKey pubkey;
+            if (keys.size()) {
+                pubkey = DecodeSecret(keys[0].get_str()).GetPubKey();
+            }
+            if (pubKeys.size()) {
+                std::vector<unsigned char> vData(ParseHex(pubKeys[0].get_str()));
+                CPubKey pubkey_temp(vData.begin(), vData.end());
+                if (pubkey.size() && pubkey_temp != pubkey) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key does not match public key for address");
                 }
+                pubkey = pubkey_temp;
+            }
+            if (pubkey.size()) {
```

Comment:
Instead of checking pubkey size, check validity? (e.g. use `IsValid()` or `IsFullyValid()`)

Reply:
IsValid() just performs the same size check anyway, but yep I'll do that for the sake of clarity, I don't think we need an IsFullyValid() check here